Add Agentic Coding Tile (#25)
Ada-Developers-Academy/independent-ai-topics: agentic-coding/workflows.md

# Recommended Workflows

> In the previous lessons we established what agents are and how their context windows work. Now we can put those ideas to use by looking at how to structure a coding session: what phases we move through, where human review belongs in that sequence, and what patterns experienced practitioners reach for when the work scales up.

Reviewer: ```suggestion
In the previous lessons we established what agents are and how their context windows work. Now we can put those ideas to use by looking at how to structure a coding session: what phases we move through, where human review belongs in that sequence, and what patterns experienced practitioners reach for when work scales up.
```

## The Core Workflow

There is no single universal workflow for agentic coding, but practitioners across the industry tend to converge on the same basic shape: 
1. plan first
2. implement next
3. then review & iterate until we reach a stopping point

It's important to note that, though we explicitly mention reviewing in the last phase, it is not the only place where reviews are necessary.
- If we as humans don't review and approve an implementation plan before sending it to an agent to write code, then that planning step loses much of its purpose. 

We'll talk a bit more about what it means to keep a human in the loop, then dive into the workflow phases in more detail!

### Human-in-the-Loop

The phrase "human-in-the-loop" refers to the principle that human judgment should sit at defined checkpoints in any agentic workflow, rather than allowing the agent to run autonomously from prompt to completion.

> This matters for the same reasons we've brought up previously: agents operate on statistical text prediction, not judgment. The outputs they produce can be sophisticated and superficially correct while containing subtle errors, missing edge cases, or violating architectural constraints that weren't fully specified. Because the AI loop runs without human awareness between steps, problems can compound before we have a chance to catch them if we do not build explicit checkpoints into our flows.

Reviewer: I suggested to change "Because" to "When" so there's more clarity on the condition this happens (when humans aren't in the loop). When reading, to me, "Because" communicated that it would happen regardless of humans being in the loop. 
```suggestion
This matters for the same reasons we've brought up previously: agents operate on statistical text prediction, not judgment. The outputs they produce can be sophisticated and superficially correct while containing subtle errors, missing edge cases, or violating architectural constraints that weren't fully specified. When the AI loop runs without human awareness between steps, problems can compound before we have a chance to catch them if we do not build explicit checkpoints into our flows.
```

The act of writing outputs to disk at these checkpoints also makes them durable across sessions, which connects back to what we've learned about managing context and ensuring important decisions persist.

### Phase 1: Planning

The planning phase establishes what we're building and how, before any code is written. This is standard practice for writing code in development environments whether or not AI is involved. The clearer our requirements and constraints are ahead of time, the less likely we are to lose time following leads that don't meet requirements or match our existing patterns.

This matters more with AI agents than it does when handing a coding task off to another human. Agents aren't able to reason about a project and may not ask questions for clarification if something in the design doesn't make sense or conflicts with itself. 
- Once an agent begins generating code, it is working from statistical patterns over the context it has been given. A vague or incomplete plan leads to implementations that have to be substantially revised.

A strong planning session typically involves two distinct activities: **research** and **specification**.

#### Research

> During research, an agent gathers the information it will need to work *well*. This might be reading through existing codebase structure, reviewing relevant documentation, understanding the constraints or edge cases involved, or exploring multiple possible approaches before committing to one. 

Reviewer: Curious to why there is emphasis on "well"? Not sure that I am seeing its importance.

Author: I think the emphasis on "well" was meant to imply what we've touched on previously: that without the research, the AI will be producing more generic work based solely on it's training data. 

How does the follow sound to better reflect that intent? 

```suggestion
During research, an agent gathers the information it will need to augment it's default training data. This might be reading through existing codebase structure, reviewing relevant documentation, discovering the constraints or edge cases involved, or exploring multiple possible approaches before committing to one. 
```

Reviewer: Gotcha! And yeah, I like this change. I can clearly see that now!

Research is a context-window-intensive activity. As we covered previously, exploration fills context windows quickly. Keeping research in its own focused session or delegating it to a subagent prevents research overhead from degrading the context available for implementation.

Artifacts from research might look like a mapping of the code base or information about structural patterns that will influence the specification. 
- To avoid burning through tokens on repeated research tasks, these artifacts are typically useful to save to disk so that future agents can read them when relevant.

#### Specification 

The "specification" portion of planning means producing a concrete, written description of what we're going to build: 
- the requirements
- the approach
- the architecture
- the testing strategy

> - any other constraints or guidelines

Reviewer: ```suggestion
- Requirements
- Approach
- Architecture
- Testing strategy
- Any other constraints or guidelines
```

Writing this to a file rather than keeping it in the conversation serves two purposes: 

> 1. It makes the plan durable: it persists beyond any single session and can be shared with colleagues for review. 

Author: ```suggestion
1. It makes the plan durable: it persists beyond any single session and any future AI agent can read that file to work from the same plan.  
```

> 2. It gives us a concrete artifact to review, share with our team, and push back on if necessary before any code is written. A vague agreement is much easier to spot when it's written down as a specific plan vs. when it is buried in a context window.

Reviewer: To me, the second half of bullet point 1 ("...can be shared with colleagues for review.") feels like it what bullet point 2 is talking about. What if we removed that section that way point 1 is solely about durability and point 2 solely about shareability?

Author: How do you feel about the suggestion added above?

Reviewer: Nice, that works for me!

> The goal of the planning phase is not to produce an acceptable specification on the first pass. One more time, because this is so vital to take with us: the intent of a planning phase is not to ask an AI to create a plan and accept the first or even second implementation plan presented to us. The purpose of a planning phase is to produce a plan we are genuinely confident in after scrutiny. This is the phase where we should ask the hardest questions: What are the edge cases? What would break this approach? Is there a simpler mechanism? 

Reviewer: How would we feel about adding some type of stylization to text to emphasize it? 
```suggestion
The goal of the planning phase is not to produce an acceptable specification on the first pass. One more time, because this is so vital to take with us: *the intent of a planning phase is not to ask an AI to create a plan and accept the first or even second implementation plan presented to us. The purpose of a planning phase is to produce a plan we are genuinely confident in after scrutiny.* This is the phase where we should ask the hardest questions: What are the edge cases? What would break this approach? Is there a simpler mechanism? 
```

When the implementation has reached the criteria defined in the plan, or when we have hit a defined stopping point, we move to review.

### !callout-info

## Try it out!

Now let's carry the plan we saved earlier into an implementation session.

1. **Point the agent to the saved plan file.** Rather than re-explaining the task, reference the plan by its file path so the agent works from the reviewed specification instead of a fresh, unreviewed description.
2. **Include testing in the same session.** Ask the agent to write and run tests as part of implementing each piece, not as a separate step tackled afterward. This keeps the relevant context available while it's still useful.
3. **Commit at logical checkpoints.** As distinct pieces of the plan are completed and tests pass, commit that work before moving to the next piece. These commits give us recovery points if a later change causes problems.
4. **Watch for repeated, unproductive cycling.** It's normal for an agent to run a test, hit a failure, and try again. If we notice the same failing approach repeating without progress, pause the session and redirect it, either by adding missing context or narrowing the current step.

> By the end of this, we should have working code committed in stages, produced from a plan we already reviewed (rather than one improvised along the way).

Reviewer: Maybe adding a mention in here about the code being tested to reinforce that the code should be tested in this phase?

Author: How does this read?
```suggestion
By the end of these steps we should have tested and working code that has been committed in stages and produced from a plan we already reviewed (rather than one improvised along the way).
```

Reviewer: Exactly what I was thinking!
